### PR TITLE
Introduce kernel module with basic types

### PR DESCRIPTION
**Pull request by TheMMaciek**

### CODE REVIEW

File: modules/kernel/src/main/scala/org/tesselation/kernel/Cell.scala [hunk 1]
```diff
@@ -0,0 +1,74 @@
+package org.tesselation.kernel
+
+import cats._
+import cats.implicits._
```
Comment (marcinwadon) [inline modules/kernel/src/main/scala/org/tesselation/kernel/Cell.scala:4]: Please use new syntax here, `import cats.implicits._` should not be used anymore, instead:
`import cats.syntax.all._` or `import cats.syntax.functor._` and so on for every feature you want

File: modules/kernel/src/main/scala/org/tesselation/kernel/PipeArrow.scala [hunk 2]
```diff
@@ -0,0 +1,21 @@
+package org.tesselation.kernel
+
+import cats.arrow.Arrow
+import cats.effect.IO
+
+import fs2.{Pipe, Stream}
+
+object PipeArrow {
+  implicit val arrowInstance: Arrow[Pipe[IO, *, *]] = new Arrow[Pipe[IO, *, *]] {
```
Comment (marcinwadon) [inline modules/kernel/src/main/scala/org/tesselation/kernel/PipeArrow.scala:9]: I'd name it `arrowInstanceForIO` or `arrowIOInstance`

File: modules/kernel/src/main/scala/org/tesselation/kernel/Stack.scala [hunk 3]
```diff
@@ -0,0 +1,38 @@
+package org.tesselation.kernel
+
+import cats.syntax.all._
```
Comment (marcinwadon) [inline modules/kernel/src/main/scala/org/tesselation/kernel/Stack.scala:3]: If possible, please import explicitly what's needed.

File: modules/kernel/src/test/scala/org/tesselation/kernel/CellMonoidTest.scala [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+import cats.effect.IO
+import cats.syntax.all._
+
+import org.tesselation.kernel.Cell.NullTerminal
+
+import higherkindness.droste.{AlgebraM, CoalgebraM, scheme}
+import weaver.SimpleIOSuite
+import weaver.scalacheck.Checkers
+
+object CellMonoidTest extends SimpleIOSuite with Checkers {
```
Comment (marcinwadon) [inline modules/kernel/src/test/scala/org/tesselation/kernel/CellMonoidTest.scala:13]: imo it should be tested using Discipline or cats-laws

**Review by marcinwadon — CHANGES_REQUESTED** [review]

File: modules/kernel/src/main/scala/org/tesselation/kernel/Cell.scala (hunk 1 above)
Comment (TheMMaciek) [reply modules/kernel/src/main/scala/org/tesselation/kernel/Cell.scala:4]: Right!

File: modules/kernel/src/main/scala/org/tesselation/kernel/Stack.scala (hunk 3 above)
Comment (TheMMaciek) [reply modules/kernel/src/main/scala/org/tesselation/kernel/Stack.scala:3]: Ok, so we are changing the approach right? So far we were importing everything.

File: modules/kernel/src/test/scala/org/tesselation/kernel/CellMonoidLawsSuite.scala [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+  // WARN: By comparing cell's run method result we can consider monoid for cells fulfilling the monoid laws.
+  //       Because two functions with the same implementation are never equal to each other we can't compare
+  //       the cells to each other by comparing the cell objects (neither convert nor hylo functions will be ever equal).
+  implicit val eq: Eq[Cell[IO, StackF, Ω, Either[CellError, Ω], Ω]] =
+    (x: Cell[IO, StackF, Ω, Either[CellError, Ω], Ω], y: Cell[IO, StackF, Ω, Either[CellError, Ω], Ω]) => {
+      for {
+        xResult <- x.run()
+        yResult <- y.run()
+      } yield xResult == yResult
+    }.unsafeRunSync()
```
Comment (marcinwadon) [inline modules/kernel/src/test/scala/org/tesselation/kernel/CellMonoidLawsSuite.scala:27]: IMO we should write it without this unsafe run sync

**Review by marcinwadon — COMMENTED** [review]
Looks solid :+1:

File: modules/kernel/src/test/scala/org/tesselation/kernel/CellMonoidLawsSuite.scala (hunk 5 above)
Comment (TheMMaciek) [reply modules/kernel/src/test/scala/org/tesselation/kernel/CellMonoidLawsSuite.scala:27]: Maybe I could test it with a different monad 🤔

Comment (TheMMaciek) [reply modules/kernel/src/test/scala/org/tesselation/kernel/CellMonoidLawsSuite.scala:27]: Switched to Id, will push the update in a second.

**Review by marcinwadon — APPROVED** [review]

**Pull request merged**